### ISSUE TITLE
363 - Deleting and Adding residues/segments

### ISSUE THREAD

**Issue by richardjgowers** [MDAnalysis/mdanalysis#985]
So I've got moving between existing residues and segments in #967 , but I need to check how adding/deleting residues/segments should work. (everything below is written as atom/residue but is equally valid for residue/segment).
### Adding

If I set `atomgroup.resids` to a new value, this should create a new Residue.

Internally, this needs to add a row to all Residue properties (resname etc).

| Res index | Resid | Resname |
| --- | --- | --- |
| 0 | 7 | 'A' |
| 1 | 8 | 'B' |
| 2 | 9 | 'C' |

``` python
ag = u.select_atoms('resid 7')
ag[:4].resids = 2
```

| Res index | Resid | Resname |
| --- | --- | --- |
| 0 | 7 | 'A' |
| 1 | 8 | 'B' |
| 2 | 9 | 'C' |
| 3 | 2 | '???' |

With the new Residue, should this take values from the old residue it split from? or should it take some default placeholder value?

This will then change `u.residues` to be size+1
### Deleting

If I empty a Residue (set all of Residue 2 contents to Residue 1) this should delete the Residue (following the old behaviour).

| Res index | Resid | Resname |
| --- | --- | --- |
| 0 | 7 | 'A' |
| 1 | 8 | 'B' |
| 2 | 9 | 'C' |

``` python
ag = u.select_atoms('resid 7')
ag.resids = 8
```

| Res index | Resid | Resname |
| --- | --- | --- |
| 0 | 7 | 'A' |
| ???1 | 8 | 'B' |
| ???2 | 9 | 'C' |

Deleting a row might cause some issues because of how we use indices.. but I'll see how it works out.

This will then change `u.residues` to be size - 1
### tl;dr

Is this above all sane and what is expected to happen?

**Comment by orbeckst** [comment]
First of apologies if the following ignores long discussions that were already had on #363 – feel free to tell me that all of this has been decided already.

Some naive questions: In #363,

1.  is *res index* a unique residue identifier (index into the residue tables)?
2.  is *resid* a label taken from the topology (equivalent what is currently *resnum*)?

Comments:
* With the new Residue, should this take values from the old residue it split from? or should it take some default placeholder value?
  * Can you set the new value to `None`, I *think* setting to undefined values is better than guessing. (I assume, `None` is not going to be a good value because you are using a string array.) In PDB files, one would use `"UNK"` as the "unknown residue name".
* deleting
  * I don't understand the example `ag = u.select_atoms('resid 7'); ag.resids = 2` – how is this deleting anything in your example, doesn't it renumber the resid 7 to resid 2?
  * There seem to be two cases (see also #984) with different intentions
     1. join two residues (eg make a residue out of resid 7 and resid 8 and give it resid 7)
     2. keep two separate residues but each with the same resid (they might be in different segments). 

We need to be clear on what *resid* means. If it is a label then we should treat it as such and just change the label, nothing more (no underlying topology changes).

Any changes in topology should then be done only using *res index* because that's the only way to sensibly express per-residue operations (instead of guessing what the user might want to happen with *resid* labels.)

**Comment by richardjgowers** [comment]
@orbeckst made a mistake in the example, it should have been `ag.resids = 8` which merges two Residues.

**Comment by dotsdl** [comment]
@richardjgowers since `resids` are just a value/label like any other attribute of a residue such as `masses`, `charges`, etc., I think it's a mistake to make it so that atom membership can be changed by using `resids`. It comes down to there being no guarantees built into the system that `resids` are unique. Instead, we should require use of `resindices` since these are the unique identifiers for a residues.

**Comment by dotsdl** [comment]
> If I set atomgroup.resids to a new value, this should create a new Residue.

You should get a `NotImplementedError` at the moment for doing this, and under the new scheme it should be impossible to set `resids` from an `AtomGroup` *by design*. This is because `resids` are a label and are a property of each residue, and so doing `ag.resids = <some_value>` should be interpreted as trying to alter the `resids` of the corresponding residues, which can't be done at the atom level (two atoms from the same residue might be setting a different resid for it, in the case `<some_value>` is an array). We should avoid this entirely IMHO unless we want to re-engage with the old problems the old topology system had.

**Comment by richardjgowers** [comment]
@dotsdl yeah, after talking to Oliver about resids (they're not unique keys!) I think you're right.  But then rather than making people use the abstract resindex, why not just let them assign using the Residue object?

```python
# splitting
res1 = u.residues[1]
res2 = u.add_Residue()  # special method which adds to topology tables etc?
res2.resid = 72
res2.resname = 'Steve'
res1[:2].residues = res2

# merging
res1, res2 = u.residues[[1, 2]]

res1.atoms.residues = res2
```

But then in the above merging example, what happens to the variable `res1`?  Does it still belong to the original Universe?

I'm going to work on other things for 363 for a while so we can nail down how we want topology manipulation to work.

**Comment by dotsdl** [comment]
@richardjgowers I think something like this is the better route. I have some thoughts on this that I think will work well that I'll try to articulate shortly. Unfortunately I've been slow to respond on all development fronts because I have a thesis to write, or something. O.o

**Comment by jbarnoud** [comment]
(I may have messed with the new project view and moved this task around by accident. Was it in progress? Or did I moved it there by mistake?)

**Comment by dotsdl** [comment]
> But then in the above merging example, what happens to the variable res1? Does it still belong to the original Universe?

@richardjgowers in your example `res1.atoms.residues = res2` shouldn't be interpreted as a merging of residues, but instead a reassignment of the atoms that were members of `res1` to membership of `res2`. After this line `res1` still exists, but it has no atoms as members. This isn't a problem.

**Comment by dotsdl** [comment]
Some thoughts, not entirely well-ordered.

1. Adding atoms to a `Topology`, though possible under the new system for `Topology` itself, is a problem for a `Universe` with an attached trajectory, since most (all?) trajectories are fixed in their atom numbers. If you're adding atoms to a system, though, I don't think there's any set of circumstances in which you'll also be iterating through trajectory data simultaneously, so not really a problem we should concern ourselves with. The same concern is already present with, e.g. `mda.Merge()`, so probably not worth thinking about further.
2. Residues and Segments are fundamentally just grouping schemes, and residues with no atoms and segments with no residues in a `Topology` are allowed. This is no problem for topology writers, either, since most topologies are just atom lists, with membership in residues and segments treated as properties of the atom entries themselves.
3. Given that empty residues and segments aren't really an issue, and given that deleting entries in the membership tables at any level (atom, residue, or segment) will result in stale `AtomGroup`, `ResidueGroup`, or `SegmentGroup` instances, we shouldn't build in mechanisms for deletion of these from the `Topology`, but only addition. If a user wants to remove things from a `Topology`, they can select out what they want from it and make a new one out of it, so we should build a mechanism (or a method) to e.g. `AtomGroup` such as `get_topology()` that extracts a `Topology` object from that selection alone. This should actually be pretty easy, as it would select all elements from each attribute of the existing `Topology` corresponding to the `AtomGroup`'s indices to build the new `Topology` object's tables.

Does this help? Is there something I've missed here?

**Comment by richardjgowers** [comment]
@dotsdl , yeah that helps

1 - Yeah adding Atom would be mostly for topology building.  There is the case of Monte Carlo... but I'm going to think about that after I've got MD sorted!

2 - Yep, Residues and Segments are easier!

3 - Yeah it is probably best just to forever expand and never delete.

**Comment by aphagstrom** [comment]
How might I delete atoms from a trajectory? I have two trajectories that have different atom numbers and I want to make them the same.

**Comment by orbeckst** [comment]
@aphagstrom This is a good question for the mailing list. As a general rule, we don't answer questions in issue reports.
